### PR TITLE
Simplify algorithm

### PR DESCRIPTION
**Pull request by sholladay**
The main goal here is to improve readability and clarity. Seems like it should perform better as well, by a minuscule amount.

### CODE REVIEW

**Comment by sholladay** [comment]
An alternative sync implementation that avoids the `while (true)` and passes the tests, but feels messy to me:

```js
module.exports.sync = function (filename, opts) {
	opts = opts || {};

	var dir = path.resolve(opts.cwd || '');

	var exists = false;
	var fp;
	while (!exists) {
		fp = path.join(dir, filename);
		exists = pathExists.sync(fp);
		if (dir === path.sep) {
			return exists ? fp : null;
		}
		dir = path.dirname(dir);
	}
	return fp;
};
```

**Comment by sholladay** [comment]
I am also happy to upgrade to `const`, remove `pinkie-promise`, etc. if there is desire for it.

**Comment by sindresorhus** [comment]
I appreciate the pull request, but I have to say, I'm not super excited about changing something that works for something slightly more readable that might have edge-cases, but I'm open to it.

-

> Seems like it should perform better as well, but I have not benchmarked it.

Doesn't matter either way. The slow part are `pathExists` calls.

File: index.js [hunk 1]
```diff
@@ -3,30 +3,22 @@ var path = require('path');
...
-			var fp = join(parts, filename);
-
+			var fp = path.join(dir, filename);
 			pathExists(fp).then(function (exists) {
 				if (exists) {
 					resolve(fp);
-				} else if (parts.pop()) {
-					find();
-				} else {
+				} else if (dir === path.sep) {
```
Comment (sindresorhus) [inline index.js]: This will cause a stack overflow on Windows when the file doesn't exist, as the root dir is something like `C:/`, not `path.sep`.

**Comment by sindresorhus** [comment]
> might have edge-cases

Make that `might`=>`will` :p https://github.com/sindresorhus/find-up/pull/7/files#r74439952

File: index.js (hunk 1 above)
Comment (sholladay) [inline index.js]: I was wondering about that. Seems like Node doesn't have an API to get the root of a path. I guess I would `substring()` to the first `path.sep` from the resolved `cwd`.

**Comment by sholladay** [comment]
> Doesn't matter either way. The slow part are pathExists calls.

Totally. Just pointing it out for the micro-optimizers out there.

**Comment by sholladay** [comment]
Okay, this works on my OS X and in CI. Needs Windows testing, though.

File: index.js [hunk 2]
```diff
@@ -3,30 +3,28 @@ var path = require('path');
...
 
-function splitPath(x) {
-	return path.resolve(x || '').split(path.sep);
-}
-
-function join(parts, filename) {
-	return path.resolve(parts.join(path.sep) + path.sep, filename);
+function rootDir(dir) {
+	var sepIndex = dir.indexOf(path.sep);
+	return sepIndex >= 0 ? dir.substring(0, sepIndex + 1) : dir;
```
Comment (sindresorhus) [inline index.js]: Actually, you could use `path.parse().root`. I was planning to drop Node.js 0.10 and 0.12 support anyways.

Comment (sholladay) [inline index.js]: Sweet. Do you want the Node >=4 conversion in a separate PR? Cause I'm happy to do that.

Comment (sindresorhus) [inline index.js]: Sure, we can finish this one first.

Comment (sholladay) [inline index.js]: Updated to use `path.parse()`. Tested Node 0.12, 4, and 6, which work. And Node 0.10, which does not.

**Comment by sindresorhus** [comment]
I sat up Windows testing and it's failing with this PR: https://ci.appveyor.com/project/sindresorhus/find-up/build/job/b1ih36mg594nb9y3 @sholladay Can you take a look?

**Comment by sholladay** [comment]
From a cursory look, this seems to be a problem with the tests themselves. In fact, I'm a little confused as to why they are passing on OS X. `bazPath` is a fully resolved absolute path, but it's passing that as the `filename` argument. Further, `<project>/fixture/baz.js` shouldn't be found by `find-up` when the cwd is `<project>`.

As documented, you're supposed to use `find-up` along the lines of:

```js
findUp('baz.js')
```

But the tests are doing:

```js
findUp('/Users/sholladay/Code/personal/find-up/fixture/baz.js');
```

I suppose I could check if the `filename` is absolute and return it immediately as is. I believe that would restore the previous behavior.

However:

 - It is not clear to me the old behavior was intentional.
 - This seems like user error.
 - The benefit to overhead ratio feels off.
 - The next release is going to be a major bump anyway.

**Pull request merged**